gradle-plugin/src/main/java/com/squareup/anvil/plugin/AnvilExtension.kt

```diff
@@ -114,31 +138,92 @@ public abstract class AnvilExtension @Inject constructor(
-   * ```
+   * @see useKsp
    */
   public val useKspBackend: Property<Boolean> = objects.property(Boolean::class.java)
-    .conventionFromProperty("com.squareup.anvil.useKspBackend", false)
+    .convention(false)
 
   /**
-   * Enables the new KSP backend for Anvil _component merging_. This is an experimental feature that
-   * currently does nothing. It's a placeholder for future work.
-   *
-   * Requires [disableComponentMerging] to be `false`.
+   * @see useKsp
+   */
+  public val useKspComponentMergingBackend: Property<Boolean> = objects.property(
+    Boolean::class.java,
+  ).convention(false)
+
+  /**
+   * Enables the new [KSP](https://github.com/google/ksp) backends for Anvil. Note that this
+   * requires the KSP plugin to already be on the buildscript classpath.
    *
+   * These modes can optionally be enabled via these gradle properties:
    * ```properties
+   * com.squareup.anvil.useKspBackend=true
    * com.squareup.anvil.useKspComponentMergingBackend=true
    * ```
+   *
+   * @param contributesAndFactoryGeneration This is an experimental feature that
+   *   replaces the previous `AnalysisHandlerExtension`-based backend, which is removed in Kotlin
+   *   2.0 and struggled with incremental compilation.
+   * @param componentMerging This is an experimental feature that currently does nothing. It's a
+   *   placeholder for future work. Requires [disableComponentMerging] to be `false`.
    */
-  public val useKspComponentMergingBackend: Property<Boolean> = objects.property(
-    Boolean::class.java,
-  )
-    .conventionFromProperty("com.squareup.anvil.useKspComponentMergingBackend", false)
+  public fun useKsp(
+    contributesAndFactoryGeneration: Boolean,
+    componentMerging: Boolean = false,
+  ) {
+    useKspBackend.setDisallowChanges(contributesAndFactoryGeneration)
+    useKspComponentMergingBackend.setDisallowChanges(componentMerging)
+    // Wire KSP
+    try {
+      project.pluginManager.apply("com.google.devtools.ksp")
+    } catch (e: Exception) {
+      // KSP not on the classpath, ask them to add it
+      error(
+        "Anvil's KSP backends require the KSP plugin to be applied to the project. " +
+          "Please apply the KSP Gradle plugin ('com.google.devtools.ksp') to your buildscript " +
+          "and try again.",
+      )
+    }
+    // Add the KSP dependency to the appropriate configurations
+    // In KMP, we only add to androidJvm/jvm targets
+    val kExtension = project.kotlinExtension
+    if (kExtension is KotlinMultiplatformExtension) {
+      kExtension.targets
+        .matching { it.platformType in SUPPORTED_PLATFORMS }
+        .configureEach {
+          addKspDep(
+            "ksp${
+              it.targetName.replaceFirstChar {
+                if (it.isLowerCase()) {
+                  it.titlecase(
+                    Locale.US,
+                  )
+                } else {
+                  it.toString()
+                }
+              }
```

Comment:
Are you okay, KtLint? 😓

Reply:
I don't like everything ktfmt does but definitely found it more reliable than ktlint heh